mne/channels/channels.py

```diff
@@ -960,6 +960,159 @@ def interpolate_bads(
+
+            def _check_pos_sphere(pos):
+                d = np.linalg.norm(pos, axis=-1)
+                d_norm = np.mean(d / np.mean(d))
+                if np.abs(1.0 - d_norm) > 0.1:
+                    warn("Your spherical fit is poor; interpolation may be inaccurate.")
+
+            _check_pos_sphere(pos_from)
+            _check_pos_sphere(pos_to)
+            mapping = _make_interpolation_matrix(pos_from, pos_to, alpha=reg)
+
+        elif method == "MNE":
+            info_eeg = pick_info(self.info, picks_good_eeg)
+            # If the original info has an average EEG reference projector but
+            # the destination info does not,
+            # update info_interp via a temporary RawArray.
+            if _has_eeg_average_ref_proj(self.info) and not _has_eeg_average_ref_proj(
+                info_interp
+            ):
+                # Create dummy data: shape (n_channels, 1)
+                temp_data = np.zeros((len(info_interp["ch_names"]), 1))
+                temp_raw = RawArray(temp_data, info_interp, first_samp=0)
+                # Using the public API, add an average reference projector.
+                temp_raw.set_eeg_reference(
+                    ref_channels="average", projection=True, verbose=False
+                )
+                # Extract the updated info.
+                info_interp = temp_raw.info
+            mapping = _map_meg_or_eeg_channels(
+                info_eeg, info_interp, mode="accurate", origin=origin
+            )
+        else:
+            raise ValueError("Unsupported interpolation method.")
+
+        # Interpolate EEG data
+        data_good = self.get_data(picks=picks_good_eeg)
+        data_interp = mapping.dot(data_good)
+
+        # Create a new instance for the interpolated EEG channels
+        if isinstance(self, BaseRaw):
+            inst_interp = RawArray(data_interp, info_interp, first_samp=self.first_samp)
+        elif isinstance(self, BaseEpochs):
+            data_interp = np.transpose(data_interp, (1, 0, 2))
+            inst_interp = EpochsArray(data_interp, info_interp)
+        elif isinstance(self, Evoked):
+            inst_interp = EvokedArray(data_interp, info_interp)
+
+        # Merge only if non-EEG channels exist
+        if not non_eeg_names_ordered:
+            return inst_interp
+
+        inst_non_eeg = self.copy().pick(non_eeg_names_ordered).load_data()
+        inst_out = inst_non_eeg.add_channels([inst_interp], force_update_info=True)
+
+        # Reorder channels
+        # Insert the entire new EEG block at the position of the first EEG channel.
+        new_order = []
+        inserted = False
+        for ch in orig_names:
+            if ch in eeg_names_orig:
+                if not inserted:
+                    new_order.extend(info_interp["ch_names"])
+                    inserted = True
+                # Skip original EEG channels.
+            else:
+                new_order.append(ch)
```

Comment:
I am a bit surprised that some version of the vectorized code I proposed in https://github.com/mne-tools/mne-python/pull/13044#issuecomment-2573525216 didn't / can't work. It's better to use such vectorized operations when possible for efficiency and compactness (and often, once you get used to it, readability)